File: 백준/Gold/23843. 콘센트/콘센트.cc
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>
using namespace std;

int a[10001], s[10001];

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    sort(a, a + n);

    priority_queue<int, vector<int>, greater<> > q;
    int idx = n - 1;
    while (idx >= 0 && q.size() < m) {
        q.push(a[idx]);
        idx--;
        if (idx < 0) {
            break;
        }
    }
    for (int i = idx; i >= 0; i--) {
        int t = q.top() + a[i];
        q.pop();
        q.push(t);
    }

    int ans = 0;
    while (!q.empty()) {
        ans = q.top();
        q.pop();
    }
    cout << ans;
}

File: 백준/Gold/23843. 콘센트/콘센트.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	static int[] a = new int[10001];

	public static void main(String[] args) throws Exception {
		BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer stringTokenizer = new StringTokenizer(bufferedReader.readLine());
		int n = Integer.parseInt(stringTokenizer.nextToken());
		int m = Integer.parseInt(stringTokenizer.nextToken());
		stringTokenizer = new StringTokenizer(bufferedReader.readLine());
		for (int i = 0; i < n; i++) {
			a[i] = Integer.parseInt(stringTokenizer.nextToken());
		}

		Arrays.sort(a, 0, n);

		Queue<Integer> pq = new PriorityQueue<>();

		int idx = n - 1;
		while (idx >= 0 && pq.size() < m) {
			pq.offer(a[idx--]);
		}

		for (int i = idx; i >= 0; i--) {
			int t = pq.poll() + a[i];
			pq.offer(t);
		}

		int ans = 0;
		while (!pq.isEmpty()) {
			ans = pq.poll();
		}
		System.out.println(ans);
	}
}
